Support expanded syntax of ports in `docker stack deploy`

yongtang opened the pull request:
**- What I did**

**NOTE: The description has been updated based on https://github.com/docker/docker/pull/30476#issuecomment-275485272**

This fix tries to address the issue in #30447 where it was not possible to specify `mode=host` for ports in compose.

**- How I did it**

This fix addresses the issue by adding support for expanded ports syntax:
```
version: "3.1"
services:
  nginx:
    image: nginx
    ports:
      - mode: host
        target: 80
        published: 9005
```


**- How to verify it**

Several unit test has been added to cover the changes.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

This fix fixes #30447.

This fix is related to #29961/#30103.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

dnephin commented:
Thanks for looking into this. Unfortunately I don't think this is the right approach.

I agree we should support a "long form" syntax for this field, but I don't think it should be the csv format. The csv format is a compromise because of limitations on the CLI. For the Compose format we should use a mapping structure with keys and values:

```yaml

ports:
  - mode: host
    target: 80
    published: 9005
```

I'm about to push a PR that does this for `volumes`, it should be up later today.

Another related issue is #29193

yongtang commented:
Thanks @dnephin for the insight and help. I will spend some time to try to get the new mapping structure working. Still trying to learn the mechanism of compose schema though I think I am close.

dnephin commented:
@yongtang you might be interested in #30521 which is some cleanup of the compose loader, and https://github.com/dnephin/docker/tree/add-expanded-mount-format-to-stack-deploy which is the branch for expanded mount format (that I thought I would have finished last week).

thaJeztah commented:
Yes, I agree with @dnephin here; @yongtang should we close this PR, or were you planning to update this one with the new approach?

yongtang commented:
Thanks @dnephin and @thaJeztah for the help, and pointing me to #30521. Very useful to understand the work flow of compose schema. I will update the PR with the new mapping structure as discussed.

yongtang committed:
Add expanded port syntax to Compose schema and types.

This commit adds expanded port syntax to Compose schema and types
so that it is possible to have
```
ports:
  - mode: host
    target: 80
    published: 9005
```

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Support expanded ports in Compose loader

This commit adds support for expanded ports in Compose loader,
and add several unit tests for loading expanded port format.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

dnephin commented on cli/compose/schema/data/config_schema_v3.1.json:
```diff
@@ -167,7 +167,19 @@
+                  "target": {"type": "integer"},
+                  "published": {"type": "integer"},
+                  "protocol": {"type": "string"}
+                }
```
I'm thinking `name` should be a field as well. I'm not sure why it's not supported in the cli.

yongtang replied:
@dnephin The `Name` is not used in the swarmkit as the moment. I added the `name` to the field. It is not actively but we could add the usage in the future.

dnephin replied:
Oh, I guess if it's not being used by swarmkit, we shouldn't add it. It's strange that it's in the API types if it's not used.

yongtang replied:
@dnephin The PR has been updated with name field removed. Please take a look.

dnephin commented on cli/compose/loader/loader_test.go:
```diff
@@ -799,3 +930,89 @@ type servicesByName []types.ServiceConfig
+	}
+
+	assert.Equal(t, 1, len(config.Services))
+	assert.Equal(t, len(expected), len(config.Services[0].Ports))
```
I would remove this assertion. It's covered by the next line, and it'll be easier to see the failure if the error shows the full diff, instead of just `x != y`

yongtang replied:
Thanks. Done.

dnephin commented on cli/compose/schema/data/config_schema_v3.1.json:
```diff
@@ -167,7 +167,19 @@
+              {"type": "string"},
+              {
+                "type": "object",
+                "properties": {
```
This should probably also have `additionalProperties: false`

yongtang replied:
Thanks @dnephin. Done.

dnephin commented on cli/compose/schema/data/config_schema_v3.1.json:
```diff
@@ -167,7 +167,19 @@
+                }
+              }
+            ],
             "format": "ports"
```
I think you might need to move `format: ports` into the `type: number`, and `type: string` cases

yongtang replied:
Thanks @dnephin. Done.

dnephin commented on cli/compose/loader/loader.go:
```diff
@@ -481,6 +485,45 @@ func transformExternal(data interface{}) (interface{}, error) {
+				}
+				for _, vv := range v {
+					ports = append(ports, vv)
+				}
```
I think you can do this as `ports = append(ports, v...)` ?

(same above)

yongtang replied:
Thanks @dnephin. There was an compilation error when appending directly:
```
cli/compose/loader/loader.go:502: cannot use v (type []"github.com/docker/docker/cli/compose/types".ServicePortConfig) as type []interface {} in append
```

Now the `toServicePortConfigs(value string)` has be changed to return `[]interface{}` (instead of `[]types.ServicePortConfig`) so it is possible to append directly with the update.

dnephin reviewed (commented):
Design LGTM,

few comments

yongtang commented:
Thanks @dnephin for the review. The PR has been updated. Please take a look and let me know if there are any additional issues.

dnephin reviewed (approved):
LGTM

Thanks!

vdemeester reviewed (approved):
LGTM 🐸
@yongtang needs a rebase :sweat_smile:

yongtang commented:
Thanks @vdemeester The PR has been rebased.

thaJeztah commented:
All green and two LGTMS

thaJeztah commented:
@dnephin do the changes to the compose format have to be synced back to the docker-compose repository? Also could you help getting these changes into the docs?

dnephin commented:
Yes, and I guess so.

friism commented:
@johnstep if you have a second and if you haven't already, could you test this on Windows? @BrandonRoyal took a quick look and couldn't get it working.

cc @icecrime

sixeyed commented:
@friism this is working for Windows in 17.04. Sample compose file:

```
version: "3.1"
services:
  iis:
    image: microsoft/iis
    ports:
      - mode: host
        target: 80
        published: 80
```

Runs with `docker stack deploy --compose-file docker-compose.yml iis`.

cc @johnstep.

friism commented:
@johnstep hey, @sixeyed tested this on Windows with 17.04 (master) server and client and it worked.

@dnephin one funny thing we found (sorry if you already thought of this) is that the error message for pre 17.04 clients is really bad (this is using the compose file that @sixeyed posted above):

    services.iis.ports.0 must be a string or number

dnephin commented:
What's bad about that error message? It seems to provide the necessary information. It has the path to offending value, and the problem with the value.

friism commented:
@dnephin it seems not good if the actual solution is "upgrade to Docker 17.04"

raarts commented:
Just a note for whoever reads this:

Using the long port format requires `version: "3.2"`, not 3.1.

karthick0070 commented:
hi,

i have tried this below stack file but i can't connect from out side of the host.
version: "3.2"
services:
 web:
  image: iiswithdb
  ports:
   - 8080:8080
   - mode: host
  deploy:
   replicas: 3

Below is this Docker ps out put

> CONTAINER ID        IMAGE                                      COMMAND                   CREATED             STATUS                  PORTS                                    NAMES
> a24cee0efb74        iiswithdb:latest                           "C:\\ServiceMonitor..."   4 minutes ago       Up Less than a second   80/tcp, 8080/tcp, 0.0.0.0:37799->0/tcp   test_web.2.mnbq86t6bgifkkuycww0udqp5

thaJeztah commented:
@karthick0070 that format is incorrect; either use the "shorthand" format (just `8080:8080`) or the _long_ format (where you specify `target:` and `published:`);

```yaml
version: "3.2"
services:
 web:
  image: iiswithdb
  ports:
    - mode: host
      target: 8080
      published: 8080
  deploy:
   replicas: 3
```

karthick0070 commented:
@thaJeztah Thanks for your help... it is working fine as excepted.

DianaDai commented:
I also has this program "services.web.ports.0 must be a string or number" when run this
version: "3.2"
services:
 web:
  image: iiswithdb
  ports:
   - 8080:8080
   - mode: host
  deploy:
   replicas: 3
